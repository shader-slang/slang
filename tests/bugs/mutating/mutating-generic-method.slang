//TEST(compute):COMPARE_COMPUTE_EX:-slang -compute -shaderobj
//TEST(compute,vulkan):COMPARE_COMPUTE_EX:-vk -slang -compute -shaderobj

// Confirm that a generic method marked `[mutating]`
// produces an `inout` parameter for `this`.

interface IGenerator
{
    [mutating] int generate();
}

struct SimpleGenerator : IGenerator
{
    int state = 0;

    [mutating] int generate() { return state++; }
}


struct Reducer
{
    int state = 0;
    [mutating] void update<S : IGenerator>(inout S g)
    {
        state = (state << 8) | state;
        state = state + g.generate();
    }
}

int test(int val)
{
    SimpleGenerator g = { 0 };
    Reducer r = { val+1 };

    r.update(g);
    r.update(g);

    return r.state;
}

//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out, name outputBuffer
RWStructuredBuffer<int> outputBuffer;

[numthreads(4, 1, 1)]
void computeMain(uint3 dispatchThreadID : SV_DispatchThreadID)
{
    int tid = dispatchThreadID.x;
    int val = test(tid);
    outputBuffer[tid] = val;
}