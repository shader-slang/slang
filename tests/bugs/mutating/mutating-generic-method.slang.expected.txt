10102
20203
30304
40405
